docs/l1/fundamentals/snap_sync.md

```diff
@@ -0,0 +1,321 @@
+
+- rootHash: state_root of the block we're trying to download
+- accountHashes: List of all the account address hashes of the storage tries to serve
+- startingHash: Storage slot hash of the first to retrieve
+- limitHash: Storage slot hash after which to stop serving
+- responseBytes: Soft limit at which to stop returning data
+
+The parameters `startingHash` and `limitHash` are only read when `accountHashes` is a single account.
+
+The return is similar to the one from `GetAccountRange`, but with multiple results, one for each account provided, with the following parameters:
+
+- slots: List of list of consecutive slots from the trie (one list per account)
+    - slotHash: Hash of the storage slot key (trie path)
+    - slotData: Data content of the slot
+- proof: List of trie nodes proving the slot range
+
+From these parameters, there is a couple of difficulties that pop up.
+
+- We need to know which accounts have storage that needs to be downloaded
+- We need to know what storage root each account has to be able to verify it
+
+To solve these issues we take two action:
+- Before we download the storage slots we ensure that the state trie is in a consistent complete state. This is accomplished by doing the insertion of accounts step first and then healing the trie. If during the storage slot download the pivot becomes stale, we heal the trie again with the new pivot, to keep the trie up to date.
+- When inserting the accounts, we grab a list of all the accounts with their storage root. If the account is healed, we marked the storage root as `None`, to indicate we should check in the DB what is the state of the storage root.
+
+#### The time traveling problem
+
+During the development of snap sync we found a recurring problem, the time traveling problem. In hash based, when a node returned to a previous state, those accounts wouldn't be redownloaded, as it would just return to the previous hash which was present in the database.
+ **Any algorithm that depended on updating accounts during healing would fail**.
+The alternative is to always go to disk if there is to mark healed accounts as unreliable and go to disk to get the datum.
+This may not be a problem in path based again, but should be studied.
+
+#### Repeated Storage Roots
+
+A large amount of the accounts with storage have exactly the same storage as other accounts.[^5] As such, when we are creating tasks for download, it's important to group the tasks by storage root and not download them twice.
+
+[^5]: This may be for a variety of reasons, but the most likely is ERC20 tokens that were deployed and never used.
+
+#### Big Accounts
+
+The storage trie is very unneven distribution of the accounts sizes. Between accounts with a single or two storage slots are around 70% of all accounts with storage trie. And large accounts have more storage slots than accounts slots are present in the state accounts. As such they need to be downloaded with special consideration.
```

Comment:
I would reword it to this:

Storage trie sizes have a very uneven distribution. Around 70% of all ethereum mainnet contracts have only 1 or 2 storage slots. However, a few contracts have more storage slots than all account leaves in the entire state trie. As such, the code needs to take this pareto distribution into account to download storage tries fast.